sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -59,68 +64,533 @@ def my_pipeline(a: int = 1, b: str = "default value"):
+          raise ValueError('compiler cannot find the ' + op_name)
+        upstream_groups, downstream_groups = (
+            self._get_uncommon_ancestors(op_groups, opsgroups_groups,
+                                         upstream_op, group))
+        dependencies[downstream_groups[0]].add(upstream_groups[0])
+
+      for subgroup in group.groups:
+        _get_dependency_opsgroup(subgroup, dependencies)
+
+    _get_dependency_opsgroup(root_group, dependencies)
+
+    return dependencies
+
+  def _resolve_value_or_reference(
+      self, value_or_reference: Union[str, dsl.PipelineParam]) -> str:
+    """_resolve_value_or_reference resolves values and PipelineParams, which could be task parameters or input parameters.
+
+    Args:
+      value_or_reference: value or reference to be resolved. It could be basic
+        python types or PipelineParam
+    """
+    if isinstance(value_or_reference, dsl.PipelineParam):
+      input_name = (
+          value_or_reference.full_name
+          if value_or_reference.op_name else value_or_reference.name)
+      if type_utils.is_parameter_type(value_or_reference.param_type):
+        return "inputs.parameters['{input_name}'].{value_field}".format(
+            input_name=input_name,
+            value_field=type_utils.get_parameter_type_field_name(
+                value_or_reference.param_type))
+      else:
+        raise NotImplementedError(
+            'Use artifact as dsl.Condition operand is not implemented yet.')
+    else:
+      if isinstance(value_or_reference, str):
+        return "'{}'".format(value_or_reference)
+      else:
+        return str(value_or_reference)
+
+  def _group_to_dag_template(
+      self,
+      group: dsl.OpsGroup,
+      inputs: Dict[str, List[Tuple[dsl.PipelineParam, str]]],
+      outputs: Dict[str, List[Tuple[dsl.PipelineParam, str]]],
+      dependencies: Dict[str, List[GroupOrOp]],
+      pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+      deployment_config: pipeline_spec_pb2.PipelineDeploymentConfig,
+      root_name: str,
+  ) -> None:
+    """Generate template given an OpsGroup.
+
+    inputs, outputs, dependencies are all helper dicts.
+    """
+    group_component_name = dsl_utils.sanitize_component_name(group.name)
+
+    if group.name == root_name:
+      group_component_spec = pipeline_spec.root
+    else:
+      group_component_spec = pipeline_spec.components[group_component_name]
+
+    # Generate component inputs spec.
+    if inputs.get(group.name, None):
+      dsl_component_spec.build_component_inputs_spec(
+          group_component_spec, [param for param, _ in inputs[group.name]])
+
+    # Generate component outputs spec.
+    if outputs.get(group.name, None):
+      group_component_spec.output_definitions.CopyFrom(
+          dsl_component_spec.build_component_outputs_spec(
+              [param for param, _ in outputs[group.name]]))
+
+    # Generate task specs and component specs for the dag.
+    sub_groups = group.groups + group.ops
```

Comment:
be consistent with the word, prefer `subgroup` everywhere rather than `sub_group`